### PR TITLE
 Add spec conformance tests for boolean-literal

### PR DESCRIPTION
**Pull request by SandaruJayawardana**
Related lang issue - [#32457](https://github.com/ballerina-platform/ballerina-lang/issues/32457)

### CODE REVIEW

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 1]
```diff
@@ -0,0 +1,161 @@
+Test-Case: output
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:1]: We need to list this under `expressions/literals`

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    boolean x = true;
+    io:println(x); // @output true
+
+    x = false;
+    io:println(x); // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals with function calls.
+Labels: function-type-descriptor, boolean-type-descriptor, boolean-literal
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:15]: This should be a `function-defn`, not a `function-type-descriptor`.

Comment (hasithaa) [reply conformance/lang/values-types-and-variables/boolean-literal.balt:15]: It would be better if we can order the labels alphabetically for the maintainability.

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+public function main() {
+    testBoolean(true);
+
+    testBoolean(false);
+
+    io:println(true.toString()); // @output true
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:22]: Better to break `toString()` validation into two a separate cases.

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+Test-Case: output
+Description: Test boolean literals with function calls.
+Labels: function-type-descriptor, boolean-type-descriptor, boolean-literal
+
+public function main() {
+    testBoolean(true);
+
+    testBoolean(false);
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:20]: `// @output` part should be here.

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+function testBoolean(boolean x) {
+    io:println(x); // @output true
+                   // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals with return statement.
+Labels: function-type-descriptor, boolean-type-descriptor, return-type-descriptor, boolean-literal
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:34]: Same as line 15 comment. I guess `return-stmt` is more correct than `return-type-descriptor` here.

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    io:println(true != false); // @output true
+
+    io:println(false == false); // @output true
+
+    io:println(false != false); // @output false
+}
+
+Test-Case: output
+Description: Test assigned boolean literals.
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:75]: Should be checking equality?

Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/boolean-literal.balt:75]: Yes. will correct this.

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+    x = true;
+    if false {
+
+    } else {
+        x = false;
+    }
+    io:println(x); // @output false
+}
+
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:107]: Another place boolean literal valid is `is-expr`

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    x = <()> false; // @error 'boolean' cannot be cast to '()'
+
+    x = <string> true; // @error 'boolean' cannot be cast to 'string'
+
+    x = <string> false; // @error 'boolean' cannot be cast to 'string'
+}
+
+Test-Case: error
+Description: Test invalid boolean literals.
+Labels: boolean-type-descriptor, int-literal
```
Comment (hasithaa) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:147]: let's add `identifier` as well.

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 9]
```diff
@@ -0,0 +1,161 @@
+Test-Case: output
+Description: Test boolean type descriptor.
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:2]: Test is for the different boolean literals against the boolean type-descriptor?

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+Labels: function-type-descriptor, boolean-type-descriptor, return-type-descriptor, boolean-literal
+
+public function main() {
+    io:println(testBoolean(true)); // @output true
+
+    io:println(testBoolean(false)); // @output false
+}
+
+function testBoolean(boolean x) returns boolean {
+    return x;
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:43]: returning true and false should be part of the return statement IMO. We can write two functions with return statements like

```ballerina
    return true;
```

```ballerina
    return false;
```

File: conformance/lang/values-types-and-variables/boolean-literal.balt [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    x = <()> true; // @error 'boolean' cannot be cast to '()'
+
+    x = <()> false; // @error 'boolean' cannot be cast to '()'
+
+    x = <string> true; // @error 'boolean' cannot be cast to 'string'
+
+    x = <string> false; // @error 'boolean' cannot be cast to 'string'
+}
+
```
Comment (MaryamZi) [inline conformance/lang/values-types-and-variables/boolean-literal.balt:144]: We can write a few more tests similar to those suggested for the nil literal.

File: conformance/lang/values-types-and-variables/boolean-literal.balt (hunk 9 above)
Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/boolean-literal.balt:2]: Yes. will update the Description.

File: conformance/lang/values-types-and-variables/boolean-literal.balt (hunk 10 above)
Comment (SandaruJayawardana) [reply conformance/lang/values-types-and-variables/boolean-literal.balt:43]: will add.

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    boolean x = true;
+    io:println(x); // @output true
+
+    x = false;
+    io:println(x); // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals with toString().
+Labels: boolean
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:15]:
```suggestion
Labels: boolean, value:toString
```

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+    io:println(true != false); // @output true
+
+    io:println(false == false); // @output true
+
+    io:println(false != false); // @output false
+}
+
+Test-Case: output
+Description: Equality check of boolean literals.
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:112]: Needs to be updated?

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    io:println(x); // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals with is-expr.
+Labels: boolean, while-stmt
+
+public function main() {
+    boolean x = false;
+    while (true) {
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:181]:
```suggestion
    while true {
```

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    if false {
+
+    } else {
+        x = false;
+    }
+    io:println(x); // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals with is-expr.
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:176]: Needs to be updated?

File: conformance/lang/expressions/literals/boolean_literal.balt (hunk 13 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/boolean_literal.balt:112]: Yes. will update the Description.

File: conformance/lang/expressions/literals/boolean_literal.balt (hunk 15 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/boolean_literal.balt:176]: Yes. will update the Description.

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+
+public function main() {
+    io:println(<boolean> true); // @output true
+
+    io:println(<boolean> false); // @output false
+}
+
+Test-Case: error
+Description: Test boolean literals with invalid type-cast-expr.
+Labels: any, boolean, decimal, float, int, null, string, type-cast-expr
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:214]:
```suggestion
Labels: any, boolean, decimal, float, int, nil-type, string, type-cast-expr
```

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+Labels: boolean, raw-template-expr
+
+public function main() {
+    io:println(`${true}`); // @output true
+
+    io:println(`${false}`); // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals as member of structural constructors.
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:249]:
```suggestion
Description: Test boolean literals as members of structural constructors.
```

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+public function main() {
+    io:println([true, false]); // @output [true,false]
+
+    io:println({"c" : true, "b" : false}); // @output {"c":true,"b":false}
+}
+
+Test-Case: error
+Description: Test invalid boolean literals.
+Labels: boolean, int
```
Comment (MaryamZi) [inline conformance/lang/expressions/literals/boolean_literal.balt:260]:
```suggestion
Labels: boolean, DecimalNumber
```

File: conformance/lang/expressions/literals/boolean_literal.balt (hunk 16 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/boolean_literal.balt:214]: Yes. Will update.

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+
+public function main() {
+    io:println(true.toBalString()); // @output true
+
+    io:println(false.toBalString()); // @output false
+}
+
+Test-Case: output
+Description: Test boolean literals with function calls.
+Labels: boolean, function-call-expr
```
Comment (hasithaa) [inline conformance/lang/expressions/literals/boolean_literal.balt:35]:
```suggestion
Labels: boolean
```

This is under minimal subset.

File: conformance/lang/expressions/literals/boolean_literal.balt [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+    testBoolean(false); // @output false
+}
+
+function testBoolean(boolean x) {
+    io:println(x);
+}
+
+Test-Case: output
+Description: Test boolean literals with return statement.
+Labels: boolean, return-stmt
```
Comment (hasithaa) [inline conformance/lang/expressions/literals/boolean_literal.balt:49]:
```suggestion
Labels: boolean
```

File: conformance/lang/expressions/literals/boolean_literal.balt (hunk 19 above)
Comment (MaryamZi) [reply conformance/lang/expressions/literals/boolean_literal.balt:35]: Since we are specifically testing these here shall we include the labels even though they are part of the minimal subset?

**Review by hasithaa — APPROVED** [review]

File: conformance/lang/expressions/literals/boolean_literal.balt (hunk 19 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/literals/boolean_literal.balt:35]: Since we are specifically testing function call in this test, @MaryamZi suggested to add this label for these types of tests.
